nmsg/kafkaio.c

```diff
@@ -0,0 +1,345 @@
+/*
+ * Copyright (c) 2024 DomainTools LLC
+ * Copyright (c) 2009-2013,2016 by Farsight Security, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* Import. */
+
+#include "private.h"
+
+#ifdef HAVE_LIBRDKAFKA
+
+struct nmsg_kafka_ctx {
+	bool running;
+	char * topic_str;
+	char * broker;
+	int partition;
+	bool consumer;
+	int timeout;
+	int64_t offset;
+	rd_kafka_conf_t * config;
+	rd_kafka_t *handle;
+	rd_kafka_topic_t *topic;
+	rd_kafka_message_t *message;
+};
+
+/* Forward. */
+static bool _kafka_addr_init(nmsg_kafka_ctx_t ctx, const char *addr);
+
+static nmsg_kafka_ctx_t _kafka_init_kafka(const char *addr, bool consumer, int timeout);
+
+/* Private. */
+
+static bool
+_kafka_addr_init(nmsg_kafka_ctx_t ctx, const char * addr)
+{
+	char *pound, *at, *comma;
+	ssize_t len;
+	pound = strchr(addr, '#');
+	at = strchr(addr, '@');
+	comma = strchr(addr, ',');
+
+	if (!pound || !at)
+		return false;
+
+	len = pound - addr;
+	ctx->topic_str = my_malloc(len + 1);
+	strncpy(ctx->topic_str, addr,len);
+	ctx->topic_str[len] = '\0';
+
+	sscanf(pound + 1, "%d", &ctx->partition);
+
+	if (comma) {
+		len = comma - at - 1;
+		ctx->broker = my_malloc(len + 1);
+		strncpy(ctx->broker, at + 1, len);
+		ctx->broker[len] = '\0';
+
+		sscanf(comma + 1, "%ld", &ctx->offset);
+	} else {
+		ctx->broker = my_malloc(strlen(at));
+		strcpy(ctx->broker, at + 1);
+		ctx->offset = RD_KAFKA_OFFSET_END;
+	}
+
+	return true;
+}
+
+static nmsg_kafka_ctx_t
+_kafka_init_kafka(const char *addr, bool consumer, int timeout)
+{
+	struct nmsg_kafka_ctx * ctx;
+	char tmp[16];
+	rd_kafka_topic_conf_t *topic_conf;
+
+	ctx = my_calloc(1, sizeof(struct nmsg_kafka_ctx));
+
+	ctx->timeout = timeout;
+	ctx->consumer = consumer;
+	ctx->running = true;
+
+	if (!_kafka_addr_init(ctx, addr)) {
+		nmsg_kafka_ctx_destroy(ctx);
+		return NULL;
+	}
+
+	ctx->config = rd_kafka_conf_new();
+	if (!ctx->config) {
+		nmsg_kafka_ctx_destroy(ctx);
+		return NULL;
+	}
+
+	snprintf(tmp, sizeof(tmp), "%i", SIGIO);
+	if (!rd_kafka_conf_set(ctx->config, "internal.termination.signal", tmp, NULL, 0) != RD_KAFKA_CONF_OK ||
+		!rd_kafka_conf_set(ctx->config, "bootstrap.servers", ctx->broker, NULL, 0) != RD_KAFKA_CONF_OK ||
```

Comment:
on github.com, this code looks like `!rd(...) != RD_KAFKA_CONF_OK`
probably don't want the first !
(pattern happens twice here)